File: ir.py
from textwrap import indent

from youtube_dl import main


class Pipeline:
    def __init__(self, body) -> None:
        self.body=body
        pass
    pass

class Stage:
    pass

class IRPrinter:
    indent=0
    class VisitMapping:
        register_map=dict() # class visit method
        @classmethod
        def register(cls,cls_type):
            def _register(fn):
                cls.register_map[cls_type]=fn
                return fn
            return _register
        @classmethod
        def get_method(cls, cls_type):
            if cls.register_map.get(cls_type) is None:
                raise RuntimeError("visit func for {} if not found".format(cls_type ))
            return cls.register_map[cls_type]
    @VisitMapping.register(Pipeline)
    def visit_pipeline(self, pipeline):
        print("pipe")
        self.visit(pipeline.body)
        pass
    @VisitMapping.register(Stage)
    def visit_stage(self, stage):
        print("stage")
        pass

    def visit(self, expr):
        fn=self.VisitMapping.get_method(expr.__class__)
        fn(self,expr)

        pass


if __name__ == "__main__":
    e=Pipeline(Stage())
    IRPrinter().visit(e)
    pass
    

File: ir.rs
pub struct Expr<T> {
    node: T,
}
pub struct Pipeline {
    agent: String,

}
pub struct Stage {
    steps: String,
}

impl Pipeline {
    pub fn make(agent: String, ) -> Pipeline {
        return Pipeline { agent: agent };
    }
}
impl Stage {
    pub fn make(steps: String) -> Stage {
        return Stage { steps: steps };
    }
}

pub struct IRPrinter {
    indent: i32,
}
impl IRPrinter {
    pub fn print_indent(&self) {
        for i in 0..self.indent {
            print!(" ");
        }
    }
    pub fn inc_indent(&mut self) {
        self.indent += 1;
    }
    pub fn dec_indent(&mut self) {
        self.indent -= 1;
    }
}
trait IRVisitor {
    fn visit(&self,  p: &IRPrinter);
}

impl IRVisitor for Pipeline {
    fn visit(&self,  p: &IRPrinter) {
        p.print_indent();
        println!("Pipeline{{");
        let mp=&mut p;
        // mp.inc_indent();
        // self.indent.visit(p);
        // mp.dec_indent();
        p.print_indent();
        print!("}}\n");
    }
}

// impl IRVisitor for Stage {
//     fn visit(&self, &mut p: IRPrinter) {
//         p.print_indent();
//         println!("Stage{{");
//         p.inc_indent();
//         // print!("{}", self.agent);
//         print!("\n");
//         p.dec_indent();
//         p.print_indent();
//         print!("}}\n");
//     }
// }

fn main() {
    let e = Pipeline::make("hahaha".to_string());
    let e2 = Stage::make("hahaha".to_string());
    let p = IRPrinter { indent: 0 };
    // e.visit(p);
    // e2.visit(&p);
    // p.visit(e);
    // println!("{}", e.node.agent);
}
